### PR TITLE
[CUDA graphs] Option to maintain CUDA RNG states on the device

### PR DESCRIPTION
**Pull request by mcarilli**
Update 11/11/20:  https://github.com/pytorch/pytorch/pull/46148#issuecomment-724414794's refactor is significant enough that it's easier to start fresh PRs.
Refactor part 1: https://github.com/pytorch/pytorch/pull/47989
***

In current master, Cuda philox states are maintained+updated on the host, which is invisible to cuda graph capture.  Arslan Zulfiqar, in his experiments, wrote a heroic workaround to manually update all RNG op nodes in a captured graph after every iteration, but in principle there's no need for such headache.  If we maintain+update states on the device, all RNG ops can be captured automatically.

This PR adds an option to maintain philox states on the GPU to facilitate graph capure. The new GPU-side states are available as an alternative to CPU-side states, and CPU-side state behavior is not changed by this PR.

GPU-side states add modest latency for non-captured eager mode, so they're purely opt-in.  By default, ops use CPU-side states as they always have.  The knob to enable GPU states is
```
torch.backends.cuda._stateful_ops.state_on_device = True
```
which someone experimenting with graphs would write at the top of their script.

Right now, CPU-side state maintenance for CUDA RNG ops yields deterministic behavior, even if RNG ops run on multiple streams, as long as all the ops are enqueued from a single thread.  GPU states introduced by this PR provide the following guarantees:
1. GPU states yield identical results to CPU states, as long as cuda rng ops are enqueued on the default stream from a single thread.
2. If cuda rng ops are enqueued from a single thread onto one or more non-default streams, GPU states no longer yield identical results to CPU streams, but GPU states' results are guaranteed deterministic regardless how many streams are used or how many ops run concurrently.
3. If cuda rng ops are enqueued from multiple threads, GPU states have the same limitations as CPU states:  Numeric safety as long as state use is mutexed, but no guarantee of determinism.

Tests mostly target verification of 1.

I could add logic to the CPU path to make CPU and GPU states yield identical results even if used on multiple streams (case 2) but it would add some overhead to the (typical eager mode) CPU path.  I'm not sure if the overhead would be negligible or not.

The PR makes [dropout launch bounds](https://github.com/pytorch/pytorch/pull/46148/files#diff-ccacb105e82756aa32ba0a7351c6c500ff55eb1c3ac8906a19e0e96b447c65ceR34) less aggressive (ensuring kernels can achieve at least 50% occupancy rather than 100%).  This allows kernels to handle the new PhiloxCudaState argument struct without register spilling, but the occupancy reduction causes a 1-2% observed slowdown in dropout kernels.

The PR also relaxes launch bounds for rng kernels that used CUDA_tensor_apply2 (`poisson_cuda_kernel`, `gamma_cuda_kernel`, `bernoulli_tensor_cuda_kernel`).  I need this to help the case for my diffs, because

- In current master, the (512, 4) launch bounds on CUDA_tensor_apply2 force 32 registers per thread and cause register spilling which makes poisson_cuda_kernel's performance terrible (about 100 GB/s effective BW on my machine)
- My diffs caused additional register spilling, regressing poisson kernel about 1.5X compared to master.
- After relaxing the launch bounds, there's no more register spilling, poisson gets >3X speedup over master, even with my other diffs.

### CODE REVIEW

**Comment by pbelevich** [comment]
@mcarilli thank you for working on this! I just want to mention that this PR should not break https://github.com/pytorch/csprng

**Comment by ngimel** [comment]
@pbelevich are there some tests to make sure csprng is ok?

**Comment by pbelevich** [comment]
@ngimel @mcarilli the only thing torchcsprng has is nightly builds with current pytorch master. There is no automatic way to check if this branch/pr is compatible with torchcsprng or not. The only solution is to build/test torchcsprng locally with local build of this branch/pr.

**Comment by pbelevich** [comment]
@ngimel @ezyang what do you think about virtual methods in `CUDAGeneratorImpl` in this pr? When I started working in torchcsprng there was a concern about adding virtual methods to generators and that's why we added `DispatchKey::CustomRNGKeyId`

**Comment by mcarilli** [comment]
> @ngimel @ezyang what do you think about virtual methods in `CUDAGeneratorImpl` in this pr? When I started working in torchcsprng there was a concern about adding virtual methods to generators and that's why we added `DispatchKey::CustomRNGKeyId`

Interesting, I didn't notice `CustomRNGKeyId` while writing the PR. 

Based on tests and comments, I think `CustomRNGKeyId` intends to allow dispatch to different RNG kernels.  That's a bigger, more invasive addition than what I was going for.  Part of this PR's goal is to use existing Cuda RNG ops, alter them as little as possible, and ensure they use the CUDAGeneratorImpl interface non-divergently whether the state actually lives on the device or the host.

But maybe I don't understand `CustomRNGKeyId`'s purpose, very curious to hear what it's for and why virtual methods in CUDAGeneratorImpl are bad.

File: torch/backends/cuda/__init__.py [hunk 1]
```diff
@@ -93,5 +93,16 @@ def __setattr__(self, name, value):
         return torch._C._set_cublas_allow_tf32(value)
 
 
+class statefulOpsModule:
+    def __getattr__(self, name):
```
[the remark this answers is not in the document]
Comment (mcarilli) [reply torch/backends/cuda/__init__.py:97]: Added [test_on_device_get_set](https://github.com/pytorch/pytorch/pull/46148/files#diff-893b1eea27352f336f4cd832919e48d721e4e90186e63400b8596db6b82e7450R3235).

File: aten/src/ATen/cuda/StatefulCUDAOpsUtils.cuh [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+#include <ATen/CUDAGeneratorImpl.h>
+#include <ATen/cuda/CUDAEvent.h>
+#include <c10/core/StreamGuard.h>
+#include <c10/cuda/CUDAGuard.h>
+
+namespace at {
+namespace cuda {
+namespace philox {
+
+ constexpr uint64_t max_kernel_threads =  (uint64_t(1) << 40);
```
[the remark this answers is not in the document]
Comment (mcarilli) [reply aten/src/ATen/cuda/StatefulCUDAOpsUtils.cuh]: hopefully resolved

**Comment by ezyang** [comment]
I think I agree that CustomRNG is not appropriate for this case, for the reasoning mcarilli put forth.

However, I have some high level questions about the approach taken here.

Generally speaking, I'd prefer for there to be less knobs. This PR adds a new knob that controls the behavior of all RNG CUDA kernels, which doubles our testing burden in these cases. So I'd like to explore if there is a way to solve your problem with something that is turned on all the way. However, to know how exactly the problem is solved, I need to know more about CUDA graph capture, and why `update_offset_kernel` solves your problem. For example, we could have fused `update_offset_kernel` into RNG kernels themselves, if all you cared about was updating this particular piece of CUDA memory. Why does this not work? Additionally, your comments about Arslan Zulfiqar's work (not posted here, so I can't see) suggest that the problem is that we hard code the ENTIRE rng state on the CUDA graphs, so that when you run the graph again, you'd get the same numbers. But it seems to me that you could tweak the kernels to, for example, query a single cuda variable that records an offset to apply to the philox offsets. In regular PyTorch this would always be zero, but for CUDA graph capture you could just roll this forward each launch (and you can find the overall offset by just looking at the CPU bound RNG state.) This seems like it would be a lot simpler than what we have here. What am I missing?

**Comment by ngimel** [comment]
@ezyang fusing `update_offset_kernel` to RNG kernels themselves is not straightforward, because all the threads in the grid (not just the block) have to see updated value. It means that full grid synchronization is required, and while that is doable, it makes the code more complicated, forces everyone who wants to write random kernels in future add those non-straightforward bits to their kernels (unless we abstract it away) and likely comes with perf penalty.

**Comment by mcarilli** [comment]
@ezyang Thanks for the thorough reply, I think I can answer all your points.

> This PR adds a new knob that controls the behavior of all RNG CUDA kernels, which doubles our testing burden in these cases.

My tests cover all the affected kernels I could find, and they're lightweight.  They run in <1.4 sec on my machine.  Happy to add more coverage if needed, but so far it doesn't seem a major burden.

> So I'd like to explore if there is a way to solve your problem with something that is turned on all the way.

Do you mean a single implementation that's always-on-by-default?  That's tough: device-side state handling (when _not_ captured) adds modest latency to eager mode.  Right now, the PR is to aid an important but experimental feature.  A high priority of mine writing the PR was not to regress eager mode in the interest of experiments.

Or do you mean a single knob should toggle all cuda graph friendliness (RNG or otherwise)?  If so, I agree:  I only plan to add _one_ knob for overall capturability of stateful ops, `torch.cuda._stateful_ops.state_on_device` (notice I didn't name it for RNG ops specifically).  Right now it only affects RNG ops, but as we discover other ops with non-capturable statefulness, I can rig their behavior to be controlled by the same flag (of course, for any such op, i'll first see if we can move state handling to the device without regressing non-captured eager mode, in which case the flag won't be relevant).

> For example, we could have fused update_offset_kernel into RNG kernels themselves, if all you cared about was updating this particular piece of CUDA memory. Why does this not work?

Very fun question :).  That's how I originally wrote it, see [an earlier version of unpack](https://github.com/pytorch/pytorch/pull/46148/files/bc109d589840f7bdbe252b9152573ec33a1b722d#diff-f80ec3b18f289e4616ed6a5de298c271ae24c4f04e11c1f0eb669ad84e58c59aR18-R36).  `PhiloxCudaState` contained a pointer to the offset value to be used by the current kernel, which all threads would read, as well as a pointer to the next offset value, which thread 0 would update.  All data and logic was buried in `PhiloxCudaState` and `unpack`, it would occur invisibly and automatically in consumer kernels with the same interface as for CPU-maintained states.  After freezing the `PhiloxCudaState` for the current kernel, I [ping-ponged the current and next offset pointers](https://github.com/pytorch/pytorch/pull/46148/files/bc109d589840f7bdbe252b9152573ec33a1b722d#diff-da1ea35f4fdb63aa9f83eed21deefd305b00f65279c7791f39f0f90bbbdc5a66R374) on the CPU, so the kernel after the current kernel would see updated offset as current and the current kernel's consumed offset as the placeholder for the updated offset thread 0 would write. I thought it was very clever and gave myself a big pat on the back. **narrator:  it was not clever.**  
<details>
  <summary>Spoiler</summary>
  The CPU-side ping ponging WAS a bit of statefulness.  After capturing, it wouldn't occur anymore when rerunning the captured graph.  If the captured region contained an odd number of RNG calls (and therefore an odd number of swaps) if I reran the captured graph then resumed non-captured eager mode ops, the next eager mode RNG kernel would use the wrong offset pointer, a subtle silent numerical bug.

I thought hard how to solve this problem while keeping the update fused into RNG consumer kernels and avoiding race conditions.  There are ways (e.g. use a single offset pointer and have thread 0 of the last block do the update, as in the cuda [threadfence sum() example](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#memory-fence-functions)) but I decided a separate 1-2 usec update kernel, which was much simpler conceptually and codewise, was a decent option.  The complexity of an in-kernel solution might even introduce more overhead than that.

[Here](https://github.com/pytorch/pytorch/pull/46148/files/bc109d589840f7bdbe252b9152573ec33a1b722d..4f86718e6d94d06a639c1cb37e50d12a0874ea3c#diff-f80ec3b18f289e4616ed6a5de298c271ae24c4f04e11c1f0eb669ad84e58c59aR23-R25) are the diffs where I changed the behavior of `unpack`.
</details>

**Comment by mcarilli** [comment]
@ngimel Thanks, i think we were typing at the same time :P
> forces everyone who wants to write random kernels in future add those non-straightforward bits to their kernels (unless we abstract it away)
In theory I could abstract it away, in `unpack`, but still complex.

> likely comes with perf penalty
Probably.  like i said, best concept I came up with was a single offset pointer updated by the last block to exit (as in the venerable threadfence sum example).  The atomic memory traffic would be minor overhead, but the 1 usec separate update kernel is also minor overhead, between the two I chose the separate kernel since it's much simpler.

**Comment by ezyang** [comment]
@mcarilli @ngimel and I had a call last Friday where I explained my scheme in more detail, and convinced mcarilli and ngimel that it should work. So we're waiting on a redo of this patch for the new design.

For those playing the home game, here are what I think the important points were:
- Treat an entire captured CUDA graph like an "operator". That means it knows how much Philox indices it consumes, so to launch a graph you increment the CUDA generator by the amount associated with the graph
- When recording, you capture the Philox indices in the kernel parameters at the time they were run. These indices need to be offset when you rerun the captured graph, so you wire them up so that they also (optionally?) read from some "global" GPU memory that is normally zero but can be reset for graph runs. (It's not actually global, this should be an extra argument passed at kernel launch).
- The captured CUDA graph launch first runs a little kernel to set the global gpu memory to whatever the offset should be for this run. This doesn't involve a synchronize at all, since the offset can be transmitted via kernel launch parameter.
- ngimel and I agreed that it would be OK to first merge a no-op PR that wires up the kernels so that they can be run in this way, even if there's no payoff

**Pull request closed without merging**